### PR TITLE
Whitespace tokenization

### PR DESCRIPTION
**Pull request by jayaddison**
As [discussed](https://github.com/MichaelAquilina/hashedindex/pull/16#issuecomment-621890496) in #16, it can be useful to enable the output of tokens for whitespace elements during string tokenization.

This changeset implements support for optional whitespace token generation.

### CODE REVIEW

File: hashedindex/textparser.py [hunk 1]
```diff
@@ -58,9 +58,26 @@ def normalize_unicode(text):
         return text
 
 
+def match_tokens(text, tokenize_whitespace):
+    first = True
+    for token in re.findall(_re_token, text):
+        if tokenize_whitespace and not first:
+            yield ' '
```
Comment (jayaddison) [inline hashedindex/textparser.py:65]: Most of the changes present in this PR are to enable this `yield` statement; they're required so that the rest of the code supports consuming from [`matched_tokens`](https://github.com/MichaelAquilina/hashedindex/pull/17/files#diff-bca7df3b022757454ef625ad3579a5b5R118) as a generator.

Alternatively it would've been possible to add the whitespace tokens in-place into the list returned by [`re.findall`](https://github.com/MichaelAquilina/hashedindex/pull/17/files#diff-bca7df3b022757454ef625ad3579a5b5R63), as suggested in [this Stackoverflow thread](https://stackoverflow.com/questions/5920643/add-an-item-between-each-item-already-in-the-list).

**Comment by jayaddison** [comment]
:hourglass_flowing_sand: The `get_ngrams` method needs a little attention here.  There seem to be ways for it to raise `StopIteration` exceptions in previously-functional calling code.

File: hashedindex/textparser.py (hunk 1 above)
Comment (jayaddison) [reply hashedindex/textparser.py:65]: :grey_question: This implementation doesn't appear to be correct - this will emit whitespace tokens between a word token and a punctuation token.

Comment (jayaddison) [reply hashedindex/textparser.py:65]: A better solution may be to allow the token regex to explictly match groups of whitespace as a token, and then emit one space character for each of those.

**Comment by jayaddison** [comment]
Ok; this is feeling more robust and ready for review now.

File: hashedindex/textparser.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ class InvalidStemmerException(Exception):
 _punctuation = _punctuation.replace('-', '')
 
 _re_punctuation = re.compile('[%s]' % re.escape(_punctuation))
-_re_token = re.compile(r'[A-z0-9]+')
+_re_token = re.compile(r'[A-z0-9]+|\s+')
```
Comment (jayaddison) [inline hashedindex/textparser.py:32]: Getting this correct in combination with #16 is tricky, and worth being very careful about since there'll be a mege conflict to resolve.

Comment (jayaddison) [reply hashedindex/textparser.py]: To prepare for this, I've applied some of the same regular expression & punctuation class naming conventions as #16.  It's not perfect but should reduce the potential for mistakes at conflict-resolution-time.

File: README.rst [hunk 3]
```diff
@@ -86,7 +86,7 @@ Tokens are wrapped within tuples due to the ability to specify any number of n-g
    [(u'life', u'is'), (u'is', u'about'), (u'about', u'making'), (u'making', u'an'), (u'an', u'impact'),
     (u'impact', u'not'), (u'not', u'making'), (u'making', u'an'), (u'an', u'income')]
 
-Take a look at the function's docstring for information on how to use ``stopwords``, specify a ``min_length`` for tokens, and configure token output using the ``ignore_numeric`` and ``retain_casing`` parameters.
+Take a look at the function's docstring for information on how to use ``stopwords``, specify a ``min_length`` for tokens, and configure token output using the ``ignore_numeric``, ``retain_casing`` and ``tokenize_whitespace`` parameters.
```
Comment (MichaelAquilina) [inline README.rst:89]: could you a document a use case for why `tokenize_whitespace` would be useful to someone?

File: README.rst [hunk 4]
```diff
@@ -88,6 +88,15 @@ Tokens are wrapped within tuples due to the ability to specify any number of n-g
...
 Take a look at the function's docstring for information on how to use ``stopwords``, specify a ``min_length`` for tokens, and configure token output using the ``ignore_numeric`` and ``retain_casing`` parameters.
 
+By default, ``word_tokenize`` omits whitespace from the output token stream; whitespaces are rarely useful to include in a document term index.
+
+If you need to tokenize text and re-assemble an output with spacing that matches the input, you may enable the ``tokenize_whitespace`` setting.
+
+.. code-block:: python
+
+    list(textparser.word_tokenize('Conventions.  May. Differ.', tokenize_whitespace=True))
+    [('conventions',), ('  ',), ('may',), (' ',), ('differ',)]
```
Comment (MichaelAquilina) [inline README.rst:98]: My initial idea was that whitespace tokens might be better off as a separate return value in order to make it obvious what is and is not considered whitespace (rather than having to analyse each element). 

However I now realise that its always going to be the case that the elements at even indexes will be the whitespaces and the elements at odd indexes will be words.

Might be worth explaining this in the README over here?

Comment (jayaddison) [reply README.rst:98]: @MichaelAquilina That seems like it would almost be true, but there's a catch :)

When tokenizing both puncuation and whitespace, it's not guaranteed that even-indexed elements will be whitespaces (I've added an additional test case to demonstrate this).

Comment (MichaelAquilina) [reply README.rst:98]: Yep that's a corner case I didn't think about. Same problem applies to returning a separate return value in that case (since you can't know the order to put the two results back together). Let's leave this as is and leave the whitespace detection to the caller. We could perhaps consider adding a function such as `is_whitespace` to help out if needs be.

**Comment by jayaddison** [comment]
Brilliant, thank you very much :)

**Comment by MichaelAquilina** [comment]
0.8.0 released!

**Comment by jayaddison** [comment]
...and now incorporated into [hashedixsearch 0.2.0](https://pypi.org/project/hashedixsearch/) :partying_face:
